### PR TITLE
Adds the debug info command

### LINKED ISSUE #2015: add odo debug info command
parent #822
debug proposal in #1995

#### `odo debug info`
Informs the user if the debug mode is enabled for the component and if the port-forward process is running or not, and information about ports used.

### PR DESCRIPTION
**Pull request by mik-dass**
**What type of PR is this?**
Adds the debug info command

/kind feature

**What does does this PR do / why we need it**:
This PR adds the debug info command. It detects if a debug session is running on the component or not.

**Which issue(s) this PR fixes**:

Fixes #2015 

**How to test changes / Special notes to the reviewer**:
- Start a debug port-forward session on a component
- run debug info on the component from a different terminal window

### CODE REVIEW

File: tests/integration/cmd_debug_test.go [hunk 1]
```diff
@@ -65,4 +65,45 @@ var _ = Describe("odo debug command tests", func() {
...
+
+	Context("odo debug info should work on a odo component", func() {
+		It("should start a debug session and run debug info on a running debug session", func() {
+			helper.CopyExample(filepath.Join("source", "nodejs"), context)
+			helper.CmdShouldPass("odo", "component", "create", "nodejs:8", "--project", project, "--context", context)
+			helper.CmdShouldPass("odo", "push", "--context", context)
+
+			stopChannel := make(chan int)
+			go func() {
+				helper.CmdShouldRunAndTerminate(60*time.Second, stopChannel, "odo", "debug", "port-forward", "--local-port", "9000", "--context", context)
```
Comment (amitkrout) [inline tests/integration/cmd_debug_test.go]: @mik-dass +100
Thanks for removing sleep()

File: tests/helper/helper_run.go [hunk 2]
```diff
@@ -42,6 +42,28 @@ func CmdShouldRunWithTimeout(timeout time.Duration, program string, args ...stri
 	return string(session.Out.Contents())
 }
 
+// CmdShouldRunAndTerminate waits and returns stdout after a closed signal is passed on the closed channel
+func CmdShouldRunAndTerminate(timeoutAfter time.Duration, stopChan chan int, program string, args ...string) string {
```
Comment (amitkrout) [inline tests/helper/helper_run.go]: +1

**Comment by amitkrout** [comment]
Hitting test failure on windows 
```
PS C:\Users\Admin\go\src\github.com\openshift\odo> make test-cmd-debug
The system cannot find the path specified.
'grep' is not recognized as an internal or external command,
operable program or batch file.
ginkgo  -randomizeAllSpecs -slowSpecThreshold=120 -timeout 7200s -nodes=1 -focus="odo debug command tests" tests/integration/
Running Suite: Integration Suite
================================
Random Seed: 1579689127 - Will randomize all specs
Will run 4 of 146 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS+SSSSSSSSSSSSSSSSSCreated dir: C:\Users\Admin\AppData\Local\Temp\453929366
Creating a new project: kvlpmtnlgn
Running odo.exe with args [odo project create kvlpmtnlgn -w -v4]
[odo]  -  Waiting for project to come up  ...
[odo] I0122 10:34:24.680721    8828 occlient.go:537] Project kvlpmtnlgn now exists
 V  Waiting for project to come up [909ms]
[odo]  V  Project 'kvlpmtnlgn' is ready for use
[odo]  V  New project created and now using project: kvlpmtnlgn
Running odo.exe with args [odo component create nodejs:8 nodejs-cmp-kvlpmtnlgn --project kvlpmtnlgn --context C:\Users\Admin\AppData\Local\Temp\453929366]
 V  Validating component [174ms]
[odo]
[odo] Please use `odo push` command to create the component with source deployed
Running odo.exe with args [odo push --context C:\Users\Admin\AppData\Local\Temp\453929366]
[odo] Validation
 V  Checking component [260ms]
[odo]
[odo] Configuration changes
[odo]  V  Initializing component
 V  Creating component [820ms]
[odo]
[odo] Pushing to component nodejs-cmp-kvlpmtnlgn of type local
 V  Checking files for pushing [97ms]
 V  Waiting for component to start [42s]
 V  Syncing files to the component [4s]
 V  Building component [9s]
[odo]  V  Changes successfully pushed to component
Checking http://localhost:9001, for WebSockets request was expected
try 0 of 12
Running odo.exe with args [odo debug port-forward --local-port 9001 --context C:\Users\Admin\AppData\Local\Temp\453929366]
[odo] Started port forwarding at ports - 9001:5858
Running odo.exe with args [odo debug info --context C:\Users\Admin\AppData\Local\Temp\453929366]
[odo] Debug is not running for the component nodejs-cmp-kvlpmtnlgn
[odo]
Deleting project: kvlpmtnlgn
Running odo.exe with args [odo project delete kvlpmtnlgn -f]
[odo] This project contains the following applications, which will be deleted
[odo] Application app
[odo] This application has following components that will be deleted
[odo] component named nodejs-cmp-kvlpmtnlgn
[odo] No services / could not get services
 V  Deleting project kvlpmtnlgn [6s]
[odo]  V  Deleted project : kvlpmtnlgn
Deleting dir: C:\Users\Admin\AppData\Local\Temp\453929366

------------------------------
+ Failure [72.161 seconds]
odo debug command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_debug_test.go:14
  odo debug info should work on a odo component
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_debug_test.go:69
    should start a debug session and run debug info on a closed debug session [It]
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_debug_test.go:89

    Expected
        <string>: Debug is not running for the component nodejs-cmp-kvlpmtnlgn


    to contain substring
        <string>: 9001

    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_debug_test.go:103
```

File: tests/integration/cmd_debug_test.go [hunk 3]
```diff
@@ -65,4 +65,47 @@ var _ = Describe("odo debug command tests", func() {
...
+			}()
+
+			// 400 response expected because the endpoint expects a websocket request and we are doing a HTTP GET
+			// We are just using this to validate if nodejs agent is listening on the other side
+			helper.HttpWaitForWithStatus("http://localhost:9000", "WebSockets request was expected", 12, 5, 400)
+			runningString := helper.CmdShouldPass("odo", "debug", "info", "--context", context)
+			Expect(runningString).To(ContainSubstring("9000"))
+			Expect(helper.ListFilesInDir(os.TempDir())).To(ContainElement(project + "-app" + "-nodejs-cmp-" + project + "-odo-debug.json"))
+			stopChannel <- 0
```
Comment (amitkrout) [inline tests/integration/cmd_debug_test.go]:
```
[odo] I0122 10:51:26.152451    6592 info.go:115] error sending signal 0 to pid 6024, 
cause: not supported by windows
```
Details -

Comment (mik-dass) [reply tests/integration/cmd_debug_test.go]: fixed

File: tests/integration/cmd_debug_test.go [hunk 4]
```diff
@@ -65,4 +66,51 @@ var _ = Describe("odo debug command tests", func() {
...
+			}()
+
+			// 400 response expected because the endpoint expects a websocket request and we are doing a HTTP GET
+			// We are just using this to validate if nodejs agent is listening on the other side
+			helper.HttpWaitForWithStatus("http://localhost:"+freePort, "WebSockets request was expected", 12, 5, 400)
+			runningString := helper.CmdShouldPass("odo", "debug", "info", "--context", context)
+			Expect(runningString).To(ContainSubstring(freePort))
+			stopChannel <- 0
+			failString := helper.CmdShouldPass("odo", "debug", "info", "--context", context)
+			Expect(failString).To(ContainSubstring("not running"))
```
Comment (amitkrout) [inline tests/integration/cmd_debug_test.go]: @mik-dass Fails here
```
Created dir: C:\Users\Admin\AppData\Local\Temp\208489730
Creating a new project: xfvetizwbz
Running odo.exe with args [odo project create xfvetizwbz -w -v4]
[odo]  -  Waiting for project to come up  ...
[odo] I0123 12:54:57.497491    4288 occlient.go:537] Project xfvetizwbz now exists
 V  Waiting for project to come up [2s]
[odo]  V  Project 'xfvetizwbz' is ready for use
[odo]  V  New project created and now using project: xfvetizwbz
Running odo.exe with args [odo component create nodejs:8 nodejs-cmp-xfvetizwbz --project xfvetizwbz --context C:\Users\Admin\AppData\Local\Temp\208489730]
 V  Validating component [158ms]
[odo]
[odo] Please use `odo push` command to create the component with source deployed
Running odo.exe with args [odo push --context C:\Users\Admin\AppData\Local\Temp\208489730]
[odo] Validation
 V  Checking component [258ms]
[odo]
[odo] Configuration changes
[odo]  V  Initializing component
 V  Creating component [741ms]
[odo]
[odo] Pushing to component nodejs-cmp-xfvetizwbz of type local
 V  Checking files for pushing [107ms]
 V  Waiting for component to start [35s]
 V  Syncing files to the component [17s]
 V  Building component [9s]
[odo]  V  Changes successfully pushed to component
Checking http://localhost:52540, for WebSockets request was expected
try 0 of 12
Running odo.exe with args [odo debug port-forward --local-port 52540 --context C:\Users\Admin\AppData\Local\Temp\208489730]
[odo] Started port forwarding at ports - 52540:5858
Running odo.exe with args [odo debug info --context C:\Users\Admin\AppData\Local\Temp\208489730]
[odo] Debug is running for the component on the local port : 52540
[odo]
Running odo.exe with args [odo debug info --context C:\Users\Admin\AppData\Local\Temp\208489730]
[odo] E0123 12:56:03.690735    9500 portforward.go:316] error copying from local connection to remote stream: read tcp6 [::1]:52540->[::1]:52548: wsarecv: An existing connection was forcibly closed by the remote host.
[odo] Debug is running for the component on the local port : 52540
[odo]
Deleting project: xfvetizwbz
Running odo.exe with args [odo project delete xfvetizwbz -f]
[odo] E0123 12:56:04.345777    9500 portforward.go:316] error copying from local connection to remote stream: read tcp6 [::1]:52540->[::1]:52551: wsarecv: An existing connection was forcibly closed by the remote host.
[odo] This project contains the following applications, which will be deleted
[odo] Application app
[odo] This application has following components that will be deleted
[odo] component named nodejs-cmp-xfvetizwbz
[odo] No services / could not get services
 V  Deleting project xfvetizwbz [6s]
[odo]  V  Deleted project : xfvetizwbz
Deleting dir: C:\Users\Admin\AppData\Local\Temp\208489730

------------------------------
+ Failure [77.365 seconds]
odo debug command tests
C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_debug_test.go:15
  odo debug info should work on a odo component
  C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_debug_test.go:70
    should start a debug session and run debug info on a closed debug session [It]
    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_debug_test.go:92

    Expected
        <string>: Debug is running for the component on the local port : 52540


    to contain substring
        <string>: not running

    C:/Users/Admin/go/src/github.com/openshift/odo/tests/integration/cmd_debug_test.go:111
```

**Comment by mik-dass** [comment]
@amitkrout I have added some comments describing the problems with windows. Please have a look.

**Comment by amitkrout** [comment]
> @amitkrout I have added some comments describing the problems with windows. Please have a look.

Thanks for the clarification. Test changes looks good to me. Can somebody from dev have look @girishramnani @kadel

**Comment by kadel** [comment]
I can't make this work :-((

I've pushed component, then I run `odo debug port-forward` in  one terminal window than in another window I executed `odo debug info` But I just get `Debug is not running for the component java-spring-boot-yfid`

File: pkg/debug/info.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	AppName        string
+	ComponentName  string
+	RemotePort     int
+	LocalPort      int
+}
+
+// GetDebugInfoFilePath gets the file path of the debug info file
+func GetDebugInfoFilePath(client *occlient.Client, componentName, appName string) string {
+	tempDir := os.TempDir()
+	debugFilePrefix := "odo-debug.json"
```
Comment (girishramnani) [inline pkg/debug/info.go]: why is this called `debugFilePrefix`? shouldn't it be called `Suffix` as it comes at the end?

File: pkg/debug/info.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+func CreateDebugInfoFile(f *DefaultPortForwarder, portPair string) error {
+	return createDebugInfoFile(f, portPair, filesystem.DefaultFs{})
+}
+
+// createDebugInfoFile creates a file in the temp directory with information regarding the debugging session of a component
+func createDebugInfoFile(f *DefaultPortForwarder, portPair string, fs filesystem.Filesystem) error {
+	portPairs := strings.Split(portPair, ":")
+	if len(portPairs) > 2 || len(portPairs) < 2 {
+		return errors.New("port pair should be of the format localPort:RemotePair")
```
Comment (girishramnani) [inline pkg/debug/info.go]: typo - `localPort:RemotePort`

File: pkg/debug/info.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func CreateDebugInfoFile(f *DefaultPortForwarder, portPair string) error {
+	return createDebugInfoFile(f, portPair, filesystem.DefaultFs{})
+}
+
+// createDebugInfoFile creates a file in the temp directory with information regarding the debugging session of a component
+func createDebugInfoFile(f *DefaultPortForwarder, portPair string, fs filesystem.Filesystem) error {
+	portPairs := strings.Split(portPair, ":")
+	if len(portPairs) > 2 || len(portPairs) < 2 {
```
Comment (girishramnani) [inline pkg/debug/info.go]: shouldn't we just check if the length is just `!=2`?

**Comment by girishramnani** [comment]
still reviewing this PR, give me till the EOD

File: pkg/debug/info.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+	var odoDebugFileData OdoDebugFile
+	err = json.Unmarshal(readFile, &odoDebugFileData)
+	if err != nil {
+		glog.V(4).Infof("couldn't unmarshal the debug file %v", debugInfoFilePath)
+		return OdoDebugFile{}, false
+	}
+
+	// get the debug process id and send a signal 0 to check if it's alive or not
+	processInfo, err := os.FindProcess(odoDebugFileData.DebugProcessId)
+	if err != nil || processInfo == nil {
```
Comment (girishramnani) [inline pkg/debug/info.go]: according to the docs of `FindProcess` https://golang.org/pkg/os/#FindProcess that
```
On Unix systems, FindProcess always succeeds and returns a Process for the given pid, regardless of whether the process exists.
```
so would this ever trigger?

Comment (mik-dass) [reply pkg/debug/info.go]: added a comment regarding this

File: pkg/debug/info.go (hunk 7 above)
Comment (mik-dass) [reply pkg/debug/info.go]: fixed

File: pkg/debug/info.go (hunk 6 above)
Comment (mik-dass) [reply pkg/debug/info.go]: fixed

File: pkg/debug/info.go (hunk 5 above)
Comment (mik-dass) [reply pkg/debug/info.go]: fixed

**Comment by mik-dass** [comment]
@girishramnani Fixed

**Comment by amitkrout** [comment]
@mik-dass Unit test is failing on travis CI - macOS. Seems your UTs changes are failing 
/hold

**Comment by amitkrout** [comment]
> @mik-dass Unit test is failing on travis CI - macOS. Seems your UTs changes are failing. I can see the same failure on my local macOS too
> /hold

@mik-dass Can you please handle #2570 in along with your pr.

**Comment by mik-dass** [comment]
> @mik-dass Can you please handle #2570 in along with your pr.

@amitkrout I will handle that in a different PR

**Comment by amitkrout** [comment]
> > @mik-dass Can you please handle #2570 in along with your pr.
> 
> @amitkrout I will handle that in a different PR

@mik-dass It is not possible this to handle in a separate pr because the unit test itself is failing due to your current pr changes. we can not merge pr with Travis CI failure. If it is a necessary requirement to separate those then you can ```remove the failure UTs from your pr only if it does not hurt the core functionality of your pr and create separate issue for it```

**Comment by mik-dass** [comment]
> @mik-dass It is not possible this to handle in a separate pr because the unit test itself is failing due to your current pr changes. we can not merge pr with Travis CI failure. If it is a necessary requirement to separate those then you can remove the failure UTs from your pr only if it does not hurt the core functionality of your pr and create separate issue for it

My PR doesn't introduce #2570 . It was introduced in https://github.com/openshift/odo/pull/2549#issuecomment-580589581. The watch failure is the same reported there. Please have a look. So I won't fix #2570 in this PR.

**Comment by amitkrout** [comment]
> > @mik-dass It is not possible this to handle in a separate pr because the unit test itself is failing due to your current pr changes. we can not merge pr with Travis CI failure. If it is a necessary requirement to separate those then you can remove the failure UTs from your pr only if it does not hurt the core functionality of your pr and create separate issue for it
> 
> My PR doesn't introduce #2570 . It was introduced in [#2549 (comment)](https://github.com/openshift/odo/pull/2549#issuecomment-580589581). The watch failure is the same reported there. Please have a look. So I won't fix #2570 in this PR.

@mik-dass Thanks for the clarification. However the current issue you are facing is nothing to deal with the error that we were getting here in https://github.com/openshift/odo/pull/2549#issuecomment-580589581. BTW the watch failure you are talking about has been fixed in the same pr just by using the latest stable golang version

Current UTs failure - 
```
ok  	github.com/openshift/odo/pkg/config	1.094s
--- FAIL: Test_getDebugInfo (0.02s)
    --- FAIL: Test_getDebugInfo/case_1:_the_debug_file_exists (0.01s)
        info_test.go:317: getDebugInfo() got = {{ } 0    0 0}, want {{OdoDebugInfo v1} 8078 testing-1 app nodejs-ex 5858 9001}
        info_test.go:320: getDebugInfo() got1 = false, want true
FAIL
FAIL	github.com/openshift/odo/pkg/debug	0.181s
```

**Comment by mik-dass** [comment]
> @mik-dass Thanks for the clarification. However the current issue you are facing is nothing to deal with the error that we were getting here in #2549 (comment). BTW the watch failure you are talking about has been fixed in the same pr just by using the latest stable golang version

I know. I reported this to you and asked you to block the PR. 

> @mik-dass Can you please handle #2570 in along with your pr.

I am saying I won't fix this flake in this PR as this has nothing to do with the current PR.

**Comment by mik-dass** [comment]
```
[odo]  ✗  Failed To Update Config To Component Deployed
```
/retest

**Comment by mik-dass** [comment]
```
ok  	github.com/openshift/odo/pkg/config	1.094s
--- FAIL: Test_getDebugInfo (0.02s)
    --- FAIL: Test_getDebugInfo/case_1:_the_debug_file_exists (0.01s)
        info_test.go:317: getDebugInfo() got = {{ } 0    0 0}, want {{OdoDebugInfo v1} 8078 testing-1 app nodejs-ex 5858 9001}
        info_test.go:320: getDebugInfo() got1 = false, want true
FAIL
FAIL	github.com/openshift/odo/pkg/debug	0.181s
```
@amitkrout fixed now.
/unhold

**Pull request merged**